File: BackEnd_Learn/Java/daily_book/src/main/java/com/dsm/daily_book/security/jwt/JwtTokenProvider.java
package com.dsm.daily_book.security.jwt;

import com.dsm.daily_book.exception.jwt.InvalidTokenException;
import com.dsm.daily_book.security.jwt.auth.AuthDetails;
import com.dsm.daily_book.security.jwt.auth.AuthDetailsService;
import io.jsonwebtoken.SignatureAlgorithm;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Component;
import io.jsonwebtoken.Jwts;

import javax.servlet.http.HttpServletRequest;
import java.util.Date;

@Component
@RequiredArgsConstructor
public class JwtTokenProvider {
    @Value("${jwt.secretkey}")
    private String SECRET_KEY;

    @Value("${jwt.prefix}")
    private String PREFIX;

    @Value("${jwt.header}")
    private String HEADER;

    @Value("${jwt.exp.access}")
    private Long accessExp;

    @Value("${jwt.exp.refresh}")
    private Long refreshExp;

    private final AuthDetailsService authDetailsService;

    public String generateAccessToken(Integer PersonalCode){
        return Jwts.builder()
                .setIssuedAt(new Date())
                .setExpiration(new Date(System.currentTimeMillis() + accessExp * 1000))
                .setSubject(PersonalCode.toString())
                .claim("type", "accesstoken")
                .signWith(SignatureAlgorithm.HS256, SECRET_KEY)
                .compact();
    }

    public String generateRefreshToken(Integer PersonalCode){
        return Jwts.builder()
                .setIssuedAt(new Date())
                .setExpiration(new Date(System.currentTimeMillis() + refreshExp * 1000))
                .setSubject(PersonalCode.toString())
                .claim("type", "refreshtoken")
                .signWith(SignatureAlgorithm.HS256, SECRET_KEY)
                .compact();
    }

    public String resolveToken(HttpServletRequest request){
        String bearer = request.getHeader(HEADER);
        if(bearer!=null && bearer.startsWith(PREFIX)){
            return bearer.substring(7);
        }
        return null;
    }

    public boolean validateToken(String token){
        try{
            Jwts.parser()
                .setSigningKey(SECRET_KEY).parseClaimsJws(token).getBody().getSubject();
            return true;
        }catch (Exception e){
            throw new InvalidTokenException();
        }
    }

    public boolean isRefreshToken(String token){
        return Jwts.parser().setSigningKey(SECRET_KEY).parseClaimsJws(token).getBody().get("type").equals("refreshtoken");
    }

    public Authentication authentication(String token){
        AuthDetails authDetails = authDetailsService.loadUserByUsername(getId(token));
        return new UsernamePasswordAuthenticationToken(authDetails, "", authDetails.getAuthorities());
    }

    public String getId(String token){
        try{
            return Jwts.parser().setSigningKey(SECRET_KEY).parseClaimsJws(token).getBody().getSubject();
        }catch(Exception e){
            throw new InvalidTokenException();
        }
    }

}

File: BackEnd_Learn/kotlin/entity/src/main/kotlin/com/example/entity/security/jwt/JwtTokenProvider.kt
package com.example.entity.security.jwt

import com.example.entity.exception.InvalidTokenException
import com.example.entity.security.jwt.auth.AuthDetails
import com.example.entity.security.jwt.auth.AuthDetailsService
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.security.oauth2.resource.OAuth2ResourceServerProperties
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component
import java.util.*
import javax.servlet.http.HttpServletRequest

@Component
class JwtTokenProvider(private val authDetailsService: AuthDetailsService) {

    @Value("\${jwt.header}")
    var header: String? = null

    @Value("\${jwt.prefix}")
    var prefix: String = ""

    @Value("\${jwt.secret}")
    var secret: String = ""

    @Value("\${jwt.exp.access}")
    var accessExp: Long = 0L

    @Value("\${jwt.exp.refresh}")
    var refreshExp: Long = 0L

    fun generateAccessToken(id: Int): String{
        return Jwts.builder()
                .signWith(SignatureAlgorithm.HS256, getSigningKey())
                .setHeaderParam("typ", "JWT")
                .setSubject(id.toString())
                .claim("type", "access")
                .setIssuedAt(Date())
                .setExpiration(Date(System.currentTimeMillis() + accessExp * 1000))
                .compact()
    }

    fun generateRefreshToken(id: Int): String{
        return Jwts.builder()
                .signWith(SignatureAlgorithm.HS256, getSigningKey())
                .setHeaderParam("typ", "JWT")
                .setSubject(id.toString())
                .claim("type", "refresh")
                .setIssuedAt(Date())
                .setExpiration(Date(System.currentTimeMillis() + refreshExp * 1000))
                .compact()
    }

    fun resolveToken(request: HttpServletRequest) : String?{
        val bearer: String= request.getHeader(header)
        if(bearer.startsWith(prefix)){
            if(bearer.length>7){
                return bearer.substring(7)
            }else throw InvalidTokenException()
        }
        return null
    }

    fun validateToken(token:String) : Boolean{
        return try{
            Jwts.parser().parseClaimsJws(token).body.subject
            true
        }catch (e: Exception){
            false
        }
    }

    fun authentication(token: String) : Authentication{
        val authDetails: AuthDetails = authDetailsService.loadUserByUsername(getId(token))
        return UsernamePasswordAuthenticationToken(authDetails, "", authDetails.authorities)
    }

    private fun getId(token: String) : String{
        return try{
            Jwts.parser()
                    .setSigningKey(getSigningKey()).parseClaimsJws(token).body.subject
        }catch (e: Exception){
            throw InvalidTokenException()
        }
    }

    private fun getSigningKey() : String{
        return Base64.getEncoder().encodeToString(secret.encodeToByteArray())
    }

}